test/test-compound-v3-vault.js

```diff
@@ -45,269 +54,489 @@ const HOUR = 3600;
+        {
+          kind: "uups",
+          unsafeAllow: ["delegatecall"],
+        }
+      );
+      await currency.connect(lp).approve(vault, MaxUint256);
+      await currency.connect(lp2).approve(vault, MaxUint256);
+      await vault.connect(admin).grantRole(getRole("LP_ROLE"), lp);
+      await vault.connect(admin).grantRole(getRole("LP_ROLE"), lp2);
+
+      return {
+        currency,
+        SingleStrategyERC4626,
+        CompoundV3InvestStrategy,
+        swapConfig,
+        vault,
+        strategy,
+        adminAddr,
+        lp,
+        lp2,
+        anon,
+        guardian,
+        admin,
+      };
+    },
+    harvestRewards: async (vault, amount) =>
+      vault.forwardToStrategy(
+        CompoundV3StrategyMethods.harvestRewards,
+        ethers.AbiCoder.defaultAbiCoder().encode(["uint256"], [amount])
+      ),
+    accessControlCheck: async (action, user, role, contract) =>
+      expect(action)
+        .to.be.revertedWithCustomError(contract, "AccessControlUnauthorizedAccount")
+        .withArgs(user, getRole(role)),
+    getSwapConfig: async (vault, strategy) => strategy.getSwapConfig(vault),
+    setSwapConfig: async (vault, swapConfig) =>
+      vault.forwardToStrategy(CompoundV3StrategyMethods.setSwapConfig, encodeSwapConfig(swapConfig)),
+  },
+];
 
-  async function setUp() {
-    const [, lp, lp2, anon, guardian, admin] = await ethers.getSigners();
-    const currency = await initForkCurrency(ADDRESSES.USDC, ADDRESSES.USDCWhale, [lp, lp2], [_A(INITIAL), _A(INITIAL)]);
+variants.forEach((variant) => {
+  describe(`${variant.name} contract tests`, function () {
+    before(async () => {
+      await setupChain(TEST_BLOCK);
+    });
 
-    const SwapLibrary = await ethers.getContractFactory("SwapLibrary");
-    const library = await SwapLibrary.deploy();
+    it("Checks vault inititializes correctly", async () => {
+      const { currency, vault, admin, anon } = await helpers.loadFixture(variant.fixture);
 
-    const CompoundV3ERC4626 = await ethers.getContractFactory("CompoundV3ERC4626", {
-      libraries: {
-        SwapLibrary: library.target,
-      },
+      expect(await vault.name()).to.equal(NAME);
+      expect(await vault.symbol()).to.equal(SYMB);
+      expect(await vault.asset()).to.equal(currency);
+      expect(await vault.totalAssets()).to.equal(0);
+      expect(await vault.hasRole(getRole("DEFAULT_ADMIN_ROLE"), admin)).to.equal(true);
+      expect(await vault.hasRole(getRole("DEFAULT_ADMIN_ROLE"), anon)).to.equal(false);
     });
-    const swapConfig = buildUniswapConfig(_W("0.00001"), FEETIER, ADDRESSES.UNISWAP);
-    const adminAddr = await ethers.resolveAddress(admin);
-    const vault = await hre.upgrades.deployProxy(CompoundV3ERC4626, [NAME, SYMB, adminAddr, swapConfig], {
-      kind: "uups",
-      constructorArgs: [ADDRESSES.cUSDCv3, ADDRESSES.REWARDS],
-      unsafeAllow: ["external-library-linking"],
+
+    if (variant.name === "CompoundV3ERC4626") {
```

Comment:
Can't this be done on a `variant.checkInitializer` function instead of special-casing it here?

Reply:
Not so easy... Not so good, since it will mean we will have tests in the variant definition.